features/storage/TESTS/kvstore/general_tests/main.cpp

```diff
@@ -0,0 +1,961 @@
+/* Copyright (c) 2017 ARM Limited
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "KVStore.h"
+#include "greentea-client/test_env.h"
+#include "unity/unity.h"
+#include "utest/utest.h"
+//#include "BlockDevice.h"
+#include "HeapBlockDevice.h"
+#include "FlashSimBlockDevice.h"
+#include "TDBStore.h"
+#include "Thread.h"
+#include "mbed_error.h"
+
+using namespace utest::v1;
+using namespace mbed;
+
+static const char   data[] = "data";
+static const size_t data_size = 5;
+static const char   key[] = "key";
+static size_t actual_size = 0;
+static char   buffer[20] = {};
+static const size_t buffer_size = 20;
+static const int num_of_threads = 3;
+
+//TEMP SETUP
+HeapBlockDevice heap_bd(4096 * 64, 1,  1, 4096);
+FlashSimBlockDevice flash_bd(&heap_bd);
+KVStore *kvstore = new TDBStore(&flash_bd);
+
+
+/*----------------initialization------------------*/
+
+//init the blockdevice
+static void kvstore_init()
+{
+    int res = kvstore->init();
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+}
+
+//deinit the blockdevice
+static void kvstore_deinit()
+{
+    int res = kvstore->deinit();
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+}
+
+/*----------------set()------------------*/
+
+//bad params : key is null
+static void set_key_null()
+{
+    int res = kvstore->set(NULL, data, data_size, 0);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : key length over key max size
+static void set_key_length_exceeds_max()
+{
+    char key_max[KVStore::MAX_KEY_SIZE + 1] = "key_name"; 
```

Comment:
This should not error. As key is a null terminated string, buffer size has no relevance. You will need to set a large string to make it error.